### PR TITLE
Update Views

### PR DESCRIPTION
**Pull request by benkeen**
This is a big update to the Views pages. Changes:

- now uses standard 2 panel layout for adding/editing/viewing
Views (includes the sidebar)
- New menu added to all Views, containing [Edit, Clone, Delete]
options. Clone and Delete are their own modals. This modal is
extensible, so can be used for different types of indexes added
via extensions.
- sidebar now properly selects views and highlights them to
show current location + view.
- Renaming a view no longer creates a second view, just renames
it. The clone feature lets you copy/clone a View. It also properly
removes the old view from the old location, be it a separate
design doc or the current one. If it removes it from another
design doc, it deletes the old design doc if it's empty.
- Auto-deleting logic of design docs has been rewritten. Before, 
we were overloading the hasViews() method on the backbone object
to determine if the design doc was now empty (this didn't 
actually work, btw). Now any code that extends Fauxton's index
types, registers the design doc property name for the index
groups (e.g. `indexes` for search indexes, `st_indexes` for geospatial
indexes or what have you). That is then used in the deletion code to 
properly identify an empty design doc.

N.B. the new menu uses React Bootstrap so it's very similar but not identical to our current menu dropdowns. The latter will be replaced in subsquent PRs.

screen shot 2016-03-07 at 8 28 55 am

### CODE REVIEW

File: app/addons/documents/index-editor/actions.js [hunk 1]
```diff
@@ -15,142 +15,282 @@ define([
...
+      }
+
+      FauxtonAPI.triggerRouteEvent('reloadDesignDocs');
 
-    updateMapCode: function (code) {
-      FauxtonAPI.dispatch({
-        type: ActionTypes.VIEW_UPDATE_MAP_CODE,
-        code: code
+      FauxtonAPI.addNotification({
+        msg: 'The <code>' + options.indexName + '</code> view has been deleted.',
```
Comment (robertkowalski) [inline app/addons/documents/index-editor/actions.js]: we need to escape `options.indexName` to make sure to open no XSS vector

something like
```js
'The <code>' + _.escape(options.indexName) + '</code> view has been deleted.',
```

File: app/addons/documents/index-editor/actions.js [hunk 2]
```diff
@@ -15,142 +15,282 @@ define([
...
 
-      designDoc.removeDdocView(viewName);
+  function deleteView (options) {
 
-      if (designDoc.hasViews()) {
-        promise = designDoc.save();
-      } else {
-        promise = designDoc.destroy();
-      }
+    var onSuccess = function () {
```
Comment (robertkowalski) [inline app/addons/documents/index-editor/actions.js]: can we make this uniform and make this a named function, too?

both ways to write a function have different scoping implications on `this`, keeping them all "named" makes it easier to scan them

File: app/addons/documents/routes-index-editor.js [hunk 3]
```diff
@@ -28,40 +28,53 @@ define([
...
+      'database/:database/new_view/:designDoc': {
+        route: 'createView',
+        roles: ['fx_loggedIn']
+      },
       'database/:database/_design/:ddoc/_view/:view': {
-        route: 'viewFn',
+        route: 'showView',
+        roles: ['fx_loggedIn']
+      },
+      'database/:database/_design/:ddoc/_view/:view/edit': {
```
Comment (robertkowalski) [inline app/addons/documents/routes-index-editor.js]: hmm.. changing the route diverges us more from the couchdb routing:

right now

```
http://localhost:8000/#/database/blerg/_design/yxc/_view/new-view


           http://localhost:5984/blerg/_design/yxc/_view/new-view
```

that makes it easy to copy / paste stuff around for debugging / development of applications that use couchdb views, i can click to the view in the fauxton ui and then copy the second part of the url 1:1 into my application code

Comment (benkeen) [inline app/addons/documents/routes-index-editor.js]: Hmm, good point. How do you suggest handling this? The edit page are separate. I guess I can keep the same URL for both, and just handle routing exclusively in the JS... but having a way to link directly to it through a URL string alone is kinda convenient. (Query string param...? Seems klutzy).

File: app/addons/documents/sidebar/sidebar.react.jsx [hunk 4]
```diff
@@ -110,38 +120,92 @@ function (app, FauxtonAPI, React, ReactDOM, Stores, Actions,
...
         );
       }, this);
     },
 
+    indexAction: function (action, params, e) {
+      e.preventDefault();
+
+      // bah. We need to close the overlay trigger whenever the user clicks on anything inside. The
+      // component doesn't have that functionality, so we simulate a click outside of it :(
+      $('body').trigger('click');
```
Comment (robertkowalski) [inline app/addons/documents/sidebar/sidebar.react.jsx]: the overlay component has an interface.

give it a `ref` and then you can call `hide` and `show` on it, example:

https://github.com/apache/couchdb-fauxton/blob/0c07ddc432c12895310592b0c46729feaac958db/app/addons/components/react-components.react.jsx#L198-L206

File: app/addons/documents/sidebar/sidebar.react.jsx [hunk 5]
```diff
@@ -110,38 +120,92 @@ function (app, FauxtonAPI, React, ReactDOM, Stores, Actions,
...
+                      Clone
+                    </li>
+                    <li onClick={this.indexAction.bind(this, 'delete', { indexName: indexName, onDelete: this.props.onDelete })}>
+                      <span className="fonticon fonticon-trash"></span>
+                      Delete
+                    </li>
+                  </ul>
+                </Popover>
+              }>
+              <span className="index-menu-toggle fonticon fonticon-wrench2"></span>
```
Comment (robertkowalski) [inline app/addons/documents/sidebar/sidebar.react.jsx]: can we make the hover-area a bit larger -- i am quite a web person but i have problems to hit the wrench icon in the sidebar from time to time

File: app/addons/documents/routes-index-editor.js (hunk 3 above)
Comment (robertkowalski) [inline app/addons/documents/routes-index-editor.js]: cool ok, makes sense

**Comment by robertkowalski** [comment]
wow! really looking forward to land this! 

:heart: :heart: :heart: 

really cool that we improved our integration test coverage - this area is one of the most important parts of fauxton, good work!

File: app/addons/documents/routes-index-editor.js (hunk 3 above)
Comment (benkeen) [inline app/addons/documents/routes-index-editor.js]: Cool thanks. I've patched the API URL to be correct for the edit page.

File: app/addons/documents/sidebar/sidebar.react.jsx (hunk 4 above)
Comment (benkeen) [inline app/addons/documents/sidebar/sidebar.react.jsx]: Good grief, you're right. I swore I tried that, honest, guv'nor!

File: app/addons/documents/sidebar/sidebar.react.jsx (hunk 5 above)
Comment (benkeen) [inline app/addons/documents/sidebar/sidebar.react.jsx]: +1 agreed

**Comment by benkeen** [comment]
Thanks, @robertkowalski! All fixed up.

**Comment by robertkowalski** [comment]
https://cloudup.com/cAczX5u4-sD

**Comment by robertkowalski** [comment]
click on existing view, choose "new document from dropdown", the form element looks misplaced:

bildschirmfoto 2016-03-11 um 10 46 27

File: app/addons/documents/tests/nightwatch/viewEdit.js [hunk 6]
```diff
@@ -12,32 +12,34 @@
...
-      .url(baseUrl + '/#/database/' + newDatabaseName + '/_design/testdesigndoc/_view/stubview')
-      .waitForElementPresent('.prettyprint', waitTime, false)
-      .assert.containsText('.prettyprint', 'stub')
+      .url(baseUrl + '/#/database/' + newDatabaseName + '/_design/testdesigndoc/_view/stubview/edit')
+      .waitForElementPresent('.index-cancel-link', waitTime, true)
+      .waitForElementNotPresent('.spinner', waitTime, true)
+      .waitForElementNotPresent('.loading-lines', waitTime, true)
+      .waitForElementVisible('#index-name', waitTime, true)
+
+      .pause(1000)
```
Comment (robertkowalski) [inline app/addons/documents/tests/nightwatch/viewEdit.js]: debug leftover?

File: app/addons/documents/index-editor/actions.js [hunk 7]
```diff
@@ -15,142 +15,287 @@ define([
...
+    var deleteDesignDoc = false;
+    if (hasRemainingIndex) {
+      promise = designDoc.save();
+    } else {
+      promise = designDoc.destroy();
+      deleteDesignDoc = true;
     }
+    promise.then(function () {
+      if (deleteDesignDoc) {
+        console.log("deleting?");
```
Comment (robertkowalski) [inline app/addons/documents/index-editor/actions.js]: debug leftover?

Comment (benkeen) [inline app/addons/documents/index-editor/actions.js]: Oops! Yup.

File: app/addons/documents/tests/nightwatch/viewEdit.js (hunk 6 above)
Comment (benkeen) [inline app/addons/documents/tests/nightwatch/viewEdit.js]: Yup. Shouldn't be there.

**Comment by benkeen** [comment]
Thanks, @robertkowalski! All fixed up. Turns out the workaround for toggling the location of the menu (top/bottom) wasn't anywhere near as hacky as I thought. It just examines the DOM node of the row to find out how close to the bottom of the page it is. Very smooth, actually.

**Comment by robertkowalski** [comment]
great work, +1000!

**Comment by benkeen** [comment]
Merged as 3ff6ff622bf22bc019a585b8285e03af94ec0650

**Pull request closed without merging**